core/templates/domain/topic_viewer/read-only-topic-object.factory.ts

```diff
@@ -0,0 +1,159 @@
+  'domain/topic/StoryReferenceObjectFactory';
+import { StorySummary } from
+  'domain/story/StorySummaryObjectFactory';
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+export interface ISkillDescriptions {
+  [skillId: string]: string | null;
+}
+
+export interface IDegreesOfMastery {
+  [skillId: string]: number | null;
+}
+
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStorySummaries: Array<StorySummary>;
+  _additionalStorySummaries: Array<StorySummary>;
+  _uncategorizedSkillSummaries: Array<SkillSummary>;
+  _subtopics: Array<Subtopic>;
+  _degreesOfMastery: IDegreesOfMastery;
+  _skillDescriptions: ISkillDescriptions;
+  _trainTabShouldBeDisplayed: boolean;
+
+  constructor(topicName: string, topicId: string,
+      canonicalStorySummaries: Array<StorySummary>,
+      additionalStorySummaries: Array<StorySummary>,
+      uncategorizedSkillSummaries: Array<SkillSummary>,
+      subtopics: Array<Subtopic>,
+      degreesOfMastery: IDegreesOfMastery,
+      skillDescriptions: ISkillDescriptions,
+      trainTabShouldBeDisplayed: boolean) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStorySummaries = canonicalStorySummaries;
+    this._additionalStorySummaries = additionalStorySummaries;
+    this._uncategorizedSkillSummaries = uncategorizedSkillSummaries;
+    this._subtopics = subtopics;
+    this._degreesOfMastery = degreesOfMastery;
+    this._skillDescriptions = skillDescriptions;
+    this._trainTabShouldBeDisplayed = trainTabShouldBeDisplayed;
+  }
+
+  getTopicName(): string {
+    return this._topicName;
+  }
+
+  getTopicId(): string {
+    return this._topicId;
+  }
+
+  getCanonicalStorySummaries(): Array<StorySummary> {
+    return this._canonicalStorySummaries.slice();
+  }
+
+  getAdditionalStorySummaries(): Array<StorySummary> {
+    return this._additionalStorySummaries.slice();
+  }
+
+  getUncategorizedSkillsSummaries(): Array<SkillSummary> {
+    return this._uncategorizedSkillSummaries.slice();
+  }
+
+  getSubtopics(): Array<Subtopic> {
+    return this._subtopics.slice();
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
+  }
+
+  getSkillDescriptions(): ISkillDescriptions {
+    return this._skillDescriptions;
+  }
+
+  getTrainTabShouldBeDisplayed(): boolean {
+    return this._trainTabShouldBeDisplayed;
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ReadOnlyTopicObjectFactory {
+  _subtopicObjectFactory: SubtopicObjectFactory;
+  _skillSummaryObjectFactory: SkillSummaryObjectFactory;
+
+  constructor(
+    private subtopicObjectFactory: SubtopicObjectFactory,
+    private skillSummaryObjectFactory: SkillSummaryObjectFactory) {
+    this._subtopicObjectFactory = subtopicObjectFactory;
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
```

Comment:
Why do we need this method? I don't see this as a part of `ReadOnlyTopicObjectFactory`. Can you explain why/where we need this?

Reply:
Actually, the `canoncialStories` list and `additionalStories` have objects of type `StorySummary`.  So to avoid repetition of code while converting the dicts to object, I had made that function.

Reply:
yeah, not a good choice. Understood where I got wrong.